### ISSUE TITLE
Bad Parsing RFC3161 (BSD)

### ISSUE THREAD

**Issue by alcampos** [syslog-ng/syslog-ng#1972]
## Version of syslog-ng

```
3.14.1
```

## Platform

Debian 8.10

# Issue

I was sending BSD messages to the syslog-ng and i noticed that the parsing was wrong. It is not parsing the PROGRAM part according the RFC. As you can see in the logs parts, the input log is

```
Apr  6 16:29:45 my.host.com {"some_json_key":"some_json_value"}
```

and when syslog-ng parse it the PROGRAM is set to **_{"some_json_key"_**. According the [RFC](https://tools.ietf.org/html/rfc3164#section-4.1.3) the PROGRAM should and i quote the RFC

> The TAG is a string of ABNF alphanumeric characters that MUST NOT exceed 32 characters.  Any non-alphanumeric character will terminate the TAG field and will be assumed to be the starting character of the CONTENT field.

In addition the RFC also specify that the PROGRAM should be like "PROGRAM: " with a space after the colon character.

> In that case, a colon and a space character usually follow the TAG. This would be displayed as "TAG: " without the quotes. 

## Configuration

```
@version: 3.14

source s_net {
	internal();
	network(ip(0.0.0.0) port(514) flags(no-parse));
};

destination d_null {
	file("/dev/stdout" template("${PROGRAM}\n"));
};


log {
	source(s_net);
	junction {
		channel { parser { syslog-parser(); }; flags(final); };
	};
	destination(d_null);
};
```

## Input and output logs
[2018-04-06T22:22:17.510078] Incoming log entry; line='Apr  6 16:29:45 my.host.com {"some_json_key":"some_json_value"}'
[2018-04-06T22:22:17.510159] Setting value; msg='0x7fe708003010', name='MESSAGE', value='Apr  6 16:29:45 my.host.com {"some_json_key":"some_json_value"}'
[2018-04-06T22:22:17.512836] Setting value; msg='0x7fe708003010', name='HOST_FROM', value='172.17.0.1'
[2018-04-06T22:22:17.512904] Setting value; msg='0x7fe708003010', name='HOST', value='172.17.0.1'
[2018-04-06T22:22:17.512936] Setting value; msg='0x7fe708003010', name='SOURCE', value='s_net'
[2018-04-06T22:22:17.513048] Message parsing complete; result='1', rule='#anon-parser0', location='/etc/syslog-ng/syslog-ng.conf:16:22'
[2018-04-06T22:22:17.514137] Outgoing message; message='{"some_json_key"\x0a'

I think in this case the PROGRAM should be empty. I would appreciate if you can point me in the right direction to fix this bug.

**Comment by bazsi** [comment]
rfc3164 is only informational and we implemented an insane amount of heuristics to our rfc3164 parser to make it work with most of the sources that we see in the wild. Not to mention that the parser in syslog-ng predates the RFC.

I am not saying we can't change that parser, but I wouldn't for this case, as I am not sure programs are not using brackets in their names.

If you are implementing a new application now, I would either recommend to use the new syslog format, or add a PROGRAM name field just for compatibility. (and extend your timestamp to have a year, timezone and fractions of a second).

btw: we have defined formats that allow json log ingestion, e.g. if the MSG part starts with the prefix "@cim:" we would parse the rest as json, as long as you are collecting these messages locally (e.g. system() driver), or with the recent default-network-drivers() source. Support for this is described in scl/cim/adapter.conf, but it's nothing but a json-parser() properly configured.

**Comment by alcampos** [comment]
Thanks for the answer. I will explain my use case. I have multiple sources and inputs to my syslog server. So i use a regex to see if the input is IETF, BSD or custom formatting, and the we parse it. In some cases we can not change the log format and i can not add PROGRAM part to the message. 

In addition i don't think PROGRAM part should contain quotes, braces or any other special character. In my experience i never saw it.

In the other hand, is there a way to validate in syslog-ng when a message is json or not? And if it json parse it?

**Comment by bazsi** [comment]
Yes, you can parse any name value pair as a json object using json-parser().

You can possible get the value parsed from the bsd message you posted using
the template $MSGHDR$MSG which should reproduce the entire text together
with the programname.

You can even fix up the values with something like:

if (program("^{")) {
   rewrite { set("$MSGHDR$MSG" value("MSG")); };

**Comment by gaborznagy** [comment]
Hi @alcampos 

- about strict PROGRAM format:
@bazsi already stated that the RFC3164 was only informational and syslog-ng has a lot of heuristics to parse correctly log messages that are not following this format.
You said that you never saw any special character in PROGRAM field. I think a good example is a typical CISCO log:
`Sep  7 06:25:23 PIXName %PIX-7-710005: UDP request discarded from 1.1.1.1/137 to outside:1.1.1.255/137`
`Mar  1 18:46:11: %SYS-5-CONFIG_I: Configured from console by vty2 (10.34.195.36)`
[source of logs](https://www.cisco.com/c/en/us/td/docs/routers/access/wireless/software/guide/SysMsgLogging.html#wp1054470)

I would recommend Bazsi's proposal to detect json messages (you said you cannot add marker in the log messages).
About the `if ()` part:
This is a new feature in syslog-ng to simplify complex configurations (e.g. channels-junctions).
You can just put Bazsi's example into you log path and that's it.

Otherwise you would have to 
1. create branches with junctions and channels, or
2. duplicate your log path and put a filter in one of them, add final flags and check the order of log path declarations. I just wrote this second example to show ugly alternatives. :)

**Comment by alcampos** [comment]
hi @gaborznagy! Thanks for the reply, both of the logs comes with a colon follow by a space. Do you think some apps really don't put the space after the colon? I think is a good practice and a good standard.

What do you think?

**Comment by bazsi** [comment]
yes, I have seen that in the wild. syslog-ng even has special support to
reproduce the non-existing space.

Bazsi

-- 
Bazsi

On Thu, Apr 12, 2018 at 12:10 AM, Alexander Campos <notifications@github.com

**Comment by gaborznagy** [comment]
I can only agree with @bazsi. And I would not change the heuristics of the RFC3164 parser either for brackets.

What is your opinion about Bazsi's workaround?
I would add that you should unset the PROGRAM macro too if you rely on it later in processing.

**Comment by alcampos** [comment]
With the @bazsi workaround i am not sure to identify BSD messages that are not compliant with the RFC and the parser. I should add a subitem in the BSD parsing to check this case and any another case. I think RFC is more than a guide, but as you comment many programs do not follow it and syslog-ng need to be prepare for the most of them.

**Comment by gaborznagy** [comment]
It is hard to filter non-RFC3164 compliant messages as they can vary in format. RFC5424 is more strict and therefore it is easier to filter using syslog-parser().
With json-parser() you can also filter messages with invalid json-format, as on parse failure the parser will drop the message.
Using the new if-else statements handling these cases should be easy.

**Comment by algernon** [comment]
I agree with @bazsi and @gaborznagy: RFC3164 was created in an attempt to formalize existing - and sometimes wildly diverging - practices. It's more of a "we've seen these common things, lets do it like that!" than a stronger specification. As such, there are many corner cases, and lots of heuristics already baked into syslog-ng that make most scenarios work. Changing them in any fundamental way - and additional heuristics about `$PROGRAM` are a fundamental change - has a fair chance of breaking something that used to work before. Which will lead to even more heuristics and even scarier code, etc. It is already far too complex for the sake of being liberal in what we accept.

Out of curiosity, in which cases can't you change the message, on the syslog-ng side, and why?

**Comment by alcampos** [comment]
Hi @algernon! Thanks for the reply too, i understand your point and i think is valid. On the syslog-ng side i can change the message but in some applications i can not change the format that is sending and the application/device is sending BSD format. I have got another example, not just json.

The following message is BSD compliance but it is bad parsed:

<14>Apr 16 14:02:51 my.host.com this,is,a,csv,and,will be bad,parsed,spam,spam,spam

[2018-04-17T20:17:13.945425] Incoming log entry; line='<14>Apr 16 14:02:51 my.host.com this,is,a,csv,and,will be bad,parsed,spam,spam,spam'
[2018-04-17T20:17:13.945493] Setting value; msg='0x7f9134002f30', name='MESSAGE', value='<14>Apr 16 14:02:51 my.host.com this,is,a,csv,and,will be bad,parsed,spam,spam,spam'
[2018-04-17T20:17:13.950315] Setting value; msg='0x7f9134002f30', name='HOST_FROM', value='172.17.0.1'
[2018-04-17T20:17:13.950578] Setting value; msg='0x7f9134002f30', name='HOST', value='172.17.0.1'
[2018-04-17T20:17:13.950613] Setting value; msg='0x7f9134002f30', name='SOURCE', value='s_net'
[2018-04-17T20:17:13.953719] Message parsing complete; result='1', rule='#anon-parser0', location='/etc/syslog-ng/syslog-ng.conf:16:22'
[2018-04-17T20:17:13.954561] Outgoing message; message='this,is,a,csv,and,will\x0a'
this,is,a,csv,and,will

 The issue here is that i have to check every message that is BSD compliance before parsing them.

Any ideas for this?

**Comment by bazsi** [comment]
hmm.. the debug messages above are from a flags(no-parse) output where
syslog-ng does not even try to parse it. what is the device that generates
this kind of messages? I can't remember anything that produced a message
without the PROGRAM name at least.

It might make sense to create a "strict" BSD parser without the heuristics,
but that needs to be applied somehow and it is not trivial how.

-- 
Bazsi

On Tue, Apr 17, 2018 at 10:19 PM, Alexander Campos <notifications@github.com

**Comment by alcampos** [comment]
This is by default the messages that produce Palo Alto Firewalls Logs. Maybe a strict BSD would be great!

**Comment by algernon** [comment]
Do all of your logs conform to RFC3161? As in, if we'd have a strict BSD parser, would that work for all your logs? Because then a combination of a network source with `no-parse` set, followed by a `parser(strict-bsd)` should work, and we'd only need to write a strict BSD parser.

**Comment by alcampos** [comment]
If i write the following:
`
junction {
	channel { parser { syslog-parser(strict-bsd); }; flags(final); };
};
`
and it works, it would be excelent!

**Comment by algernon** [comment]
I think that can be arranged, or something very much like that.
